### PR TITLE
[IMP] sale_crm: update expected_revenue based on sent quotations

### PR DESCRIPTION
**Pull request by adsh-odoo**
Purpose
=========
Help salespeople keep their opportunities up-to-date by propagating the revenues
they set on Quotations.

Specification
==============
When a quotation is confirmed, if the expected revenue of the lead is less than the
total of quote which we are confirming, the expected revenue of the lead should be
updated to match the total of the quote.

For example:

If the lead has an expected revenue of $40:

Quote 1: $45 and we confirm it then,
Then the expected revenue of the lead would be updated from $40 to $45.
Task-3806572
ENT PR https://github.com/odoo/enterprise/pull/61550

### CODE REVIEW

File: addons/sale_crm/models/sale_order.py [hunk 1]
```diff
@@ -13,3 +13,15 @@ class SaleOrder(models.Model):
 
     def action_confirm(self):
         return super(SaleOrder, self.with_context({k:v for k,v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
+
+    def message_post(self, **kwargs):
```
Comment (ibo-odoo) [inline addons/sale_crm/models/sale_order.py:17]: Hello @adsh-odoo Can't we handle this with compute method?(Make expected_revenue computed) Can you give It a try? The main drawback here is message_post will be called even If the user will log in the quotation. And this whole code will still execute even If we don't need to.
Hello @jeh-odoo What is your thoughts on this?

**Comment by adsh-odoo** [comment]
Hello @ibo-odoo 
I have updated the approach can you please have a look?
Thanks!

**Comment by adsh-odoo** [comment]
Hello @ibo-odoo 
Currently working on some new points added by Luc. So I will ping you once it's done.
Thanks

**Review by ibo-odoo — COMMENTED** [review]
Hello @adsh-odoo Can you please check the comments?

**Comment by adsh-odoo** [comment]
Hello @ibo-odoo 
Changes have been done.
Can you please have a look
Thanks !!

**Review by ibo-odoo — COMMENTED** [review]
Hello @adsh-odoo Can you please try out these comments? Thanks.

**Comment by adsh-odoo** [comment]
Hello @ibo-odoo 
Suggested changes have been made.
Can you please have a look?
Thanks !!

File: addons/sale_crm/tests/test_crm_lead_convert_quotation.py [hunk 2]
```diff
@@ -118,3 +119,56 @@ def test_lead_convert_to_quotation_nothing(self):
...
+        """
+        product1, product2 = self.env['product.template'].create([{
+            'name': 'Test product1',
+            'list_price': 100.0,
+        }, {
+            'name': 'Test product2',
+            'list_price': 200.0,
+        }])
+
+        so1 = self.env['sale.order'].create({
```
Comment (ibo-odoo) [inline addons/sale_crm/tests/test_crm_lead_convert_quotation.py:139]: Can you please check for one opportunity with multiple sale orders?

**Review by ibo-odoo — COMMENTED** [review]
Hello @adsh-odoo Same check comments.(Good to go with code but need to check test)

**Review by ibo-odoo — COMMENTED** [review]
Hello @adsh-odoo Some comments to polish the testcase. Thank you.

**Comment by adsh-odoo** [comment]
Hello @ibo-odoo 

I have made the changes can you please have a look?
Thanks!!

**Review by Odoonan — COMMENTED** [review]
@adsh-odoo 

... And the community counterpart :)

Cheers!

**Comment by adsh-odoo** [comment]
Hello @Odoonan 

I have changed the approach as per the suggestion.
Can you please have a look?

Thanks!!

File: addons/sale_crm/models/sale_order.py [hunk 3]
```diff
@@ -13,3 +13,23 @@ class SaleOrder(models.Model):
 
     def action_confirm(self):
         return super(SaleOrder, self.with_context({k:v for k,v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
+
+    def _message_post_after_hook(self, new_message, message_values):
+        self._set_expected_revenue()
```
Comment (Odoonan) [inline addons/sale_crm/models/sale_order.py:18]: Here, we only want to do something when we mark the so as sent. So I would add this condition above `self._set_expected_revenue()` : 

`if self.env.context.get('mark_so_as_sent'):`

File: addons/sale_crm/models/sale_order.py [hunk 4]
```diff
@@ -13,3 +13,23 @@ class SaleOrder(models.Model):
...
+        return super()._message_post_after_hook(new_message, message_values)
+
+    def action_cancel(self):
+        res = super().action_cancel()
+        self._set_expected_revenue()
```
Comment (Odoonan) [inline addons/sale_crm/models/sale_order.py:23]: Since the logic we add is actually modifying the lead, I would rather add a method on the crm.lead model and call it where appropriate with

`self.opportunity_id._update_expected_revenues_from_so()`

and then in the method, start with

for opportunity in self:
... UPDATE CODE ....

so that if the opp. has no lead, it still work properly. This would become the only method overriden in crm_sale_subscription.

File: addons/sale_crm/models/sale_order.py [hunk 5]
```diff
@@ -13,3 +13,23 @@ class SaleOrder(models.Model):
 
     def action_confirm(self):
         return super(SaleOrder, self.with_context({k:v for k,v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
```
Comment (Odoonan) [inline addons/sale_crm/models/sale_order.py:15]: call the method updating the code hereabove (instead of in action_cancel)

File: addons/sale_crm/models/sale_order.py [hunk 6]
```diff
@@ -13,3 +13,23 @@ class SaleOrder(models.Model):
...
+        self._set_expected_revenue()
+        return super()._message_post_after_hook(new_message, message_values)
+
+    def action_cancel(self):
+        res = super().action_cancel()
+        self._set_expected_revenue()
+        return res
+
+    def _set_expected_revenue(self):
+        if not hasattr(self, 'recurring_monthly'):
```
Comment (Odoonan) [inline addons/sale_crm/models/sale_order.py:27]: Why this?

File: addons/sale_crm/models/sale_order.py [hunk 7]
```diff
@@ -13,3 +13,23 @@ class SaleOrder(models.Model):
...
+            min_order_non_recurring_total = min([
+                order.tax_totals['amount_untaxed']
+                for order in self.opportunity_id.order_ids.filtered(lambda order: order.state != 'cancel' and order.tax_totals['amount_untaxed'])
```
Comment (Odoonan) [inline addons/sale_crm/models/sale_order.py:30]: why not simply using `order.amount_untaxed`?

File: addons/sale_crm/models/sale_order.py [hunk 8]
```diff
@@ -13,3 +13,23 @@ class SaleOrder(models.Model):
 
     def action_confirm(self):
         return super(SaleOrder, self.with_context({k:v for k,v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
+
+    def _message_post_after_hook(self, new_message, message_values):
```
Comment (Odoonan) [inline addons/sale_crm/models/sale_order.py:17]: let us keep the same signature as in sale_order:
`def _message_post_after_hook(self, message, msg_vals)`

File: addons/sale_crm/models/sale_order.py [hunk 9]
```diff
@@ -13,3 +13,23 @@ class SaleOrder(models.Model):
...
+        self._set_expected_revenue()
+        return super()._message_post_after_hook(new_message, message_values)
+
+    def action_cancel(self):
+        res = super().action_cancel()
+        self._set_expected_revenue()
+        return res
+
+    def _set_expected_revenue(self):
```
[the remark this answers is not in the document]
Comment (Odoonan) [reply addons/sale_crm/models/sale_order.py:26]: (Another way would be to check whether the field `non_recurring_total` exists on the model, and use it instead of amount_untaxed if so)

File: addons/sale_crm/models/sale_order.py [hunk 10]
```diff
@@ -13,3 +13,23 @@ class SaleOrder(models.Model):
...
+        if not hasattr(self, 'recurring_monthly'):
+            min_order_non_recurring_total = min([
+                order.tax_totals['amount_untaxed']
+                for order in self.opportunity_id.order_ids.filtered(lambda order: order.state != 'cancel' and order.tax_totals['amount_untaxed'])
+            ], default=0.0)
+
+            if self.opportunity_id.expected_revenue < min_order_non_recurring_total:
+                self.opportunity_id.expected_revenue = min_order_non_recurring_total
+                self.opportunity_id._message_log(body=_('Expected revenue is updated on the based of the linked quotes.'))
```
Comment (Odoonan) [inline addons/sale_crm/models/sale_order.py:35]: I would reword to 'has been updated based on the linked quotes'

**Review by Odoonan — COMMENTED** [review]
Hello @adsh-odoo ,

Thanks for the update, it is closer to what was discussed indeed.

Here is the follow up review.
(please apply changes in the other PR as well)

Cheers!

File: addons/sale_crm/models/sale_order.py [hunk 11]
```diff
@@ -13,3 +13,23 @@ class SaleOrder(models.Model):
...
         return super(SaleOrder, self.with_context({k:v for k,v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
+
+    def _message_post_after_hook(self, new_message, message_values):
+        self._set_expected_revenue()
+        return super()._message_post_after_hook(new_message, message_values)
+
+    def action_cancel(self):
+        res = super().action_cancel()
+        self._set_expected_revenue()
+        return res
```
[the remark this answers is not in the document]
Comment (Odoonan) [reply addons/sale_crm/models/sale_order.py:24]: ( I think we do not want any update on cancel :) )

[the remark this answers is not in the document]
Comment (adsh-odoo) [reply addons/sale_crm/models/sale_order.py:24]: If we not call our method inside `action_cancel` it will break in that case suppose amount on some quote is set as the expected revenue and we cancel that quote revenue will not be updated based on the remaining quote.
Let me know your opinion.

[the remark this answers is not in the document]
Comment (Odoonan) [reply addons/sale_crm/models/sale_order.py:24]: I think we can keep it simple, no need to handle action_cancel :)

**Comment by adsh-odoo** [comment]
Hello @Odoonan 
Suggested changes have been made can you please have a look?
Thanks!!

File: addons/sale_crm/models/sale_order.py (hunk 3 above)
Comment (adsh-odoo) [reply addons/sale_crm/models/sale_order.py]: yes, you are right otherwise `update_revenues_from_so` will be called every time while posting a certain message.

File: addons/sale_crm/models/sale_order.py (hunk 5 above)
Comment (adsh-odoo) [reply addons/sale_crm/models/sale_order.py]: Actually earlier it was not needed because `_message_post_after_hook` will be called ultimately after calling `action_confirm` but now we are using the context so we need to call our method here.

**Review by Odoonan — COMMENTED** [review]
Hello @adsh-odoo 

Thanks for the quick updates, this version is already much better :)

Small review available. Can you also include links of COM / ENT Pull requests on both branches (PR messages + commit messages) to ease navigation between PRs? Thanks!

(I still have to double check the test, but then we should be good to go)

Cheers!

File: addons/sale_crm/models/crm_lead.py [hunk 12]
```diff
@@ -101,3 +101,14 @@ def _merge_get_fields_specific(self):
         # add all the orders from all lead to merge
         fields_info['order_ids'] = lambda fname, leads: [(4, order.id) for order in leads.order_ids]
         return fields_info
+
+    def update_revenues_from_so(self):
```
Comment (Odoonan) [inline addons/sale_crm/models/crm_lead.py:105]: can we make this private ? _... same for other branch :)

File: addons/sale_crm/tests/test_sale_crm_revenue.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
+from odoo.addons.crm.tests import common as crm_common
```
Comment (Odoonan) [inline addons/sale_crm/tests/test_sale_crm_revenue.py:1]: can we import TestCrmCommon instead? :)
from odoo.addons.crm.tests.crm_common import TestCrmCommon

File: addons/sale_crm/tests/test_sale_crm_revenue.py [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+from odoo.fields import Command
+
+
+class TestSaleCrmRevenue(crm_common.TestCrmCommon):
```
Comment (Odoonan) [inline addons/sale_crm/tests/test_sale_crm_revenue.py:5]: update here

File: addons/sale_crm/tests/test_sale_crm_revenue.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    def test_sale_crm_revenue(self):
+        """ Test the updation of the expected_revenue when the quotation is sent or confirmed.
+        If the expected_revenue of the lead is smaller than the total of any quote(except cancelled quotes), update it with the lowest total.
```
Comment (Odoonan) [inline addons/sale_crm/tests/test_sale_crm_revenue.py:8]: you can use 3 lines to make it more readable

File: addons/sale_crm/tests/test_sale_crm_revenue.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        e.g. if the lead have a expected revenue of 40 $
```
Comment (Odoonan) [inline addons/sale_crm/tests/test_sale_crm_revenue.py:9]: has*

**Review by Odoonan — COMMENTED** [review]
Hello @adsh-odoo ,

Just a few comments left and we can green :)

Cheers!

File: addons/sale_crm/models/sale_order.py [hunk 17]
```diff
@@ -12,4 +12,11 @@ class SaleOrder(models.Model):
...
     def action_confirm(self):
-        return super(SaleOrder, self.with_context({k:v for k,v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
+        res = super(SaleOrder, self.with_context({k: v for k, v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
+        self.opportunity_id._update_revenues_from_so()
+        return res
+
+    def _message_post_after_hook(self, message, msg_vals):
+        if self.env.context.get('mark_so_as_sent'):
+            self.opportunity_id._update_revenues_from_so()
+        return super()._message_post_after_hook(message, msg_vals)
```
Comment (tde-banana-odoo) [inline addons/sale_crm/models/sale_order.py:22]: Posting a message is not really a business method. Not sure what kind of hook you expect here but I don't think this should be used to trigger business update.

File: addons/sale_crm/models/crm_lead.py [hunk 18]
```diff
@@ -101,3 +101,15 @@ def _merge_get_fields_specific(self):
...
         fields_info['order_ids'] = lambda fname, leads: [(4, order.id) for order in leads.order_ids]
         return fields_info
+
+    def _update_revenues_from_so(self):
+        for opportunity in self:
+            min_order_untaxed_total = min([
+                order.amount_untaxed
+                for order in opportunity.order_ids.filtered(lambda order: order.state != 'cancel' and order.amount_untaxed
+                    and self.company_id.currency_id == order.currency_id)
+            ], default=0.0)
```
Comment (tde-banana-odoo) [inline addons/sale_crm/models/crm_lead.py:111]: You should probably extract the order fetch into a sub method that can be overridden in enterprise (if necessay, still to check conditions). That way even if computation is not the same between sale_crm and sale_crm_subscription, part of the code is already common.

Comment (tde-banana-odoo) [inline addons/sale_crm/models/crm_lead.py:111]: Looks like this could benefit from a read group, to try

File: addons/sale_crm/models/crm_lead.py [hunk 19]
```diff
@@ -101,3 +101,15 @@ def _merge_get_fields_specific(self):
...
+
+            if opportunity.expected_revenue < min_order_untaxed_total:
+                opportunity.expected_revenue = min_order_untaxed_total
+                opportunity._message_log(body=_('Expected revenue has been updated based on the linked Sales orders and Quotations.'))
```
Comment (tde-banana-odoo) [inline addons/sale_crm/models/crm_lead.py:115]: This should be a tracked field. And it is ok if for some usage we don't really need the tracking, would say better have standard code than manual tracking.

Comment (tde-banana-odoo) [inline addons/sale_crm/models/crm_lead.py:115]: It is possible to filter tracking (we did it for another task), maybe you should look at ``_track_filter_for_display`` if we can do something with that

**Review by tde-banana-odoo — CHANGES_REQUESTED** [review]
Thanks for the PR ! I had a look, I think we should try to improve the implementation (optimized search, more standard tracking). Added some suggestions :) cheers !

File: addons/sale_crm/models/sale_order.py [hunk 20]
```diff
@@ -13,3 +13,23 @@ class SaleOrder(models.Model):
...
+        res = super().write(vals)
+        if vals.get('state'):
+            self.opportunity_id.with_context(tracking_disable=False)._update_revenues_from_so()
+        return res
+
+    def _filter_sales_order(self, aggregation_field):
+        domain = [
+            ('opportunity_id', '=', self.opportunity_id.id),
+            ('state', '!=', 'cancel'),
+            ('currency_id', '=', self.opportunity_id.company_id.currency_id.id),
```
Comment (Odoonan) [inline addons/sale_crm/models/sale_order.py:27]: I'm guessing this is why you do not use read_group on all opportunities at the same time?
Because we still need to only take the orders with relevant currencies?

Comment (Odoonan) [reply addons/sale_crm/models/sale_order.py:27]: (Just writing my thoughts : Maybe we could remove that part of the domain, use the read_group to return recordsets of orders per opportunity that we can filter regarding the currency, then take the min amount. But then we would still be filtering and looping on every opportunity + SO.) We'll see what TDE thinks.

Comment (Odoonan) [reply addons/sale_crm/models/sale_order.py:27]: Continuing my idea: But having a read_group called on a single opportunity each time, and grouping by opportunity, does not really make sense to me ^^

File: addons/sale_crm/models/crm_lead.py [hunk 21]
```diff
@@ -101,3 +101,10 @@ def _merge_get_fields_specific(self):
         # add all the orders from all lead to merge
         fields_info['order_ids'] = lambda fname, leads: [(4, order.id) for order in leads.order_ids]
         return fields_info
+
+    def _update_revenues_from_so(self):
+        for opportunity in self:
+            min_order_untaxed_total = opportunity.order_ids._filter_sales_order("amount_untaxed")
```
Comment (Odoonan) [inline addons/sale_crm/models/crm_lead.py:107]: It's a bit strange to use a read_group since we already call the method on the order_ids recordset.

So I'd say either we use a read_group on all opportunities in self, which would be nice if doable, or we use `_filter_sales_order` to just filter the orders with the domain presently inside of that method, and return the min value of the argument field among the filtered orders. (no read_group then)

**Review by Odoonan — COMMENTED** [review]
Hello @adsh-odoo  

Here is a pre review on last changes, since @tde-banana-odoo is not here today.

Of what I see, the best solution would be to use the read group on all orders of opportunities in the recordset, grouping by opportunity.

The only thing that could be an issue with that read_group is the currency that could be different between orders, making it impossible (afaik) to use the read group as is to return the min values of given field. Also, one of the fields is non stored.

So I would at least simplify the use of your new function to make it usable in more places, both in ENT and COM PR's.

I left other comments too.

Thanks for the quick updates, cheers! We'll get there :muscle:

File: addons/sale_crm/models/sale_order.py [hunk 22]
```diff
@@ -13,3 +13,23 @@ class SaleOrder(models.Model):
 
     def action_confirm(self):
         return super(SaleOrder, self.with_context({k:v for k,v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
+
+    def write(self, vals):
+        res = super().write(vals)
+        if vals.get('state'):
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/sale_crm/models/sale_order.py:19]: We should check the value (and that it changed), effectively

File: addons/sale_crm/models/crm_lead.py (hunk 21 above)
Comment (tde-banana-odoo) [reply addons/sale_crm/models/crm_lead.py:107]: Indeed, we have SO browsing opportunity browsing SO calling a method on SO ... Would also move the code here as pointed out by @Odoonan

File: addons/sale_crm/tests/test_sale_crm_revenue.py [hunk 23]
```diff
@@ -0,0 +1,47 @@
+from odoo.addons.crm.tests.common import TestCrmCommon
+from odoo.fields import Command
+
+
+class TestSaleCrmRevenue(TestCrmCommon):
+    def test_sale_crm_revenue(self):
```
Comment (tde-banana-odoo) [inline addons/sale_crm/tests/test_sale_crm_revenue.py:6]:
```suggestion
class TestSaleCrm(TestCrmCommon):

    def test_sale_crm_revenue(self):
```

File: addons/sale_crm/tests/test_sale_crm_revenue.py [hunk 24]
```diff
@@ -0,0 +1,47 @@
+from odoo.addons.crm.tests.common import TestCrmCommon
```
Comment (tde-banana-odoo) [inline addons/sale_crm/tests/test_sale_crm_revenue.py:1]: side note: file canbe named "test_sale_crm" as it tests the bridge itself

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for update, few comments :)

**Comment by adsh-odoo** [comment]
Hello @tde-banana-odoo @Odoonan 
Done with changes 
Can you please have a look?
Thanks!!

File: addons/sale_crm/models/sale_order.py [hunk 25]
```diff
@@ -13,3 +13,9 @@ class SaleOrder(models.Model):
 
     def action_confirm(self):
         return super(SaleOrder, self.with_context({k:v for k,v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
+
+    def write(self, vals):
+        res = super().write(vals)
+        if vals.get('state') in ['sent', 'sale', 'cancel'] and self.opportunity_id:
+            self.opportunity_id.with_context(tracking_disable=False)._update_revenues_from_so()
+        return res
```
Comment (Odoonan) [inline addons/sale_crm/models/sale_order.py:21]: I still believe we only wanted to update when reaching sent and sale states only. If we want to check the value has changed, maybe we could do this:

```python
     def write(self, vals):
        state_per_order = {}
        if vals.get('state') in ['sent', 'sale']:
            state_per_order = {
                order.id: order.state for order in self
            }
        res = super().write(vals)
        if state_per_order:
            new_state_orders = self.filtered(lambda order: order.state != state_per_order.get(order.id))
            if new_state_orders.opportunity_id:
                new_state_orders.opportunity_id.with_context(tracking_disable=False)._update_revenues_from_so()
        return res
```

File: addons/sale_crm/models/crm_lead.py [hunk 26]
```diff
@@ -101,3 +101,17 @@ def _merge_get_fields_specific(self):
...
+            [('opportunity_id', 'in', self.ids),
+             ('state', '!=', 'cancel'),
+             ('amount_untaxed', '>', 0)],
+            ['opportunity_id', 'currency_id'],
+            ['amount_untaxed:min'],
+        )
+        for opportunity, currency, min_field in data:
+            if opportunity.company_id.currency_id == currency:
+                if opportunity.expected_revenue < min_field:
+                    opportunity.expected_revenue = min_field
```
Comment (Odoonan) [inline addons/sale_crm/models/crm_lead.py:116]: since min_field is known here, let us call it `min_amount`

File: addons/sale_crm/models/sale_order.py [hunk 27]
```diff
@@ -12,4 +12,7 @@ class SaleOrder(models.Model):
         domain="[('type', '=', 'opportunity'), '|', ('company_id', '=', False), ('company_id', '=', company_id)]")
 
     def action_confirm(self):
-        return super(SaleOrder, self.with_context({k:v for k,v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
+        res = super(SaleOrder, self.with_context({k: v for k, v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
+        for order in self:
+            self.opportunity_id.with_context(order_id=order.id)._update_revenues_from_so()
```
Comment (tde-banana-odoo) [inline addons/sale_crm/models/sale_order.py:17]: You can simply give the SO as an argument, that way we don't rely on context :)

File: addons/sale_crm/models/crm_lead.py [hunk 28]
```diff
@@ -101,3 +101,14 @@ def _merge_get_fields_specific(self):
         # add all the orders from all lead to merge
         fields_info['order_ids'] = lambda fname, leads: [(4, order.id) for order in leads.order_ids]
         return fields_info
+
+    def _update_revenues_from_so(self):
```
Comment (tde-banana-odoo) [inline addons/sale_crm/models/crm_lead.py:105]:
```suggestion
    def _update_revenues_from_so(self, order):
```

File: addons/sale_crm/models/crm_lead.py [hunk 29]
```diff
@@ -101,3 +101,14 @@ def _merge_get_fields_specific(self):
...
+        for opportunity in self:
+            current_order_id = self.env['sale.order'].browse(self.env.context.get('order_id'))
+            if (
+                opportunity.expected_revenue < current_order_id.amount_untaxed
```
Comment (tde-banana-odoo) [inline addons/sale_crm/models/crm_lead.py:109]: In case people have False/null ? not sure if automatically converted to 0
```suggestion
                (opportunity.expected_revenue or 0) < current_order_id.amount_untaxed
```

File: addons/sale_crm/models/crm_lead.py [hunk 30]
```diff
@@ -101,3 +101,14 @@ def _merge_get_fields_specific(self):
...
+                and current_order_id.amount_untaxed
```
Comment (tde-banana-odoo) [inline addons/sale_crm/models/crm_lead.py:110]: this seems unnecessary (unless people have a negative number as expected revenue but hey)

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing !

**Comment by adsh-odoo** [comment]
Hello @tde-banana-odoo 
Done with the changes.
Can you please have a look?
Thanks!!

File: addons/sale_crm/models/sale_order.py [hunk 31]
```diff
@@ -12,4 +12,7 @@ class SaleOrder(models.Model):
         domain="[('type', '=', 'opportunity'), '|', ('company_id', '=', False), ('company_id', '=', company_id)]")
 
     def action_confirm(self):
-        return super(SaleOrder, self.with_context({k:v for k,v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
+        res = super(SaleOrder, self.with_context({k: v for k, v in self._context.items() if k != 'default_tag_ids'})).action_confirm()
+        for order in self:
+            self.opportunity_id._update_revenues_from_so(order)
```
Comment (tde-banana-odoo) [inline addons/sale_crm/models/sale_order.py:17]: Seems we could have a m2m-like behavior we don't want 
```suggestion
        for order in self:
            order.opportunity_id._update_revenues_from_so(order)
```

File: addons/sale_crm/models/crm_lead.py [hunk 32]
```diff
@@ -101,3 +101,12 @@ def _merge_get_fields_specific(self):
...
         return fields_info
+
+    def _update_revenues_from_so(self, order):
+        for opportunity in self:
+            if (
+                (opportunity.expected_revenue or 0) < order.amount_untaxed
+                and order.currency_id == opportunity.company_id.currency_id
+            ):
+                opportunity.expected_revenue = order.amount_untaxed
+                opportunity._track_set_log_message(_('Expected revenue has been updated based on the linked Sales orders and Quotations.'))
```
Comment (tde-banana-odoo) [inline addons/sale_crm/models/crm_lead.py:112]: We don't have quotations anymore, since only confirmed orders are taken into account now
```suggestion
                opportunity._track_set_log_message(_('Expected revenue has been updated based on the linked Sales orders.'))
```

**Comment by adsh-odoo** [comment]
Hello @tde-banana-odoo 
Changes have been made.
Can you please have a look?
Thanks!!

**Comment by tde-banana-odoo** [comment]
@robodoo r+
